packages/http/fetch/src/middlewares/compressionHandler.ts

```diff
@@ -0,0 +1,239 @@
+
+			response = await this.next?.execute(url, requestInit as RequestInit, requestOptions);
+		}
+		return response !== undefined && response !== null ? Promise.resolve(response) : Promise.reject(new Error("Response is undefined"));
+	}
+
+	private contentRangeBytesIsPresent(header: FetchHeadersInit | undefined): boolean {
+		if (!header) {
+			return false;
+		}
+		const contentRange = getRequestHeader(header, CompressionHandler.CONTENT_RANGE_HEADER);
+		return contentRange?.toLowerCase().includes("bytes") ?? false;
+	}
+
+	private contentEncodingIsPresent(header: FetchHeadersInit | undefined): boolean {
+		if (!header) {
+			return false;
+		}
+		return getRequestHeader(header, CompressionHandler.CONTENT_ENCODING_HEADER) !== undefined;
+	}
+
+	private getRequestBodySize(body: unknown): number {
+		if (!body) {
+			return 0;
+		}
+		if (typeof body === "string") {
+			return body.length;
+		}
+		if (body instanceof Blob) {
+			return body.size;
+		}
+		if (body instanceof ArrayBuffer) {
+			return body.byteLength;
+		}
+		if (ArrayBuffer.isView(body)) {
+			return body.byteLength;
+		}
+		if (inNodeEnv() && Buffer.isBuffer(body)) {
+			return body.byteLength;
+		}
+		throw new Error("Unsupported body type");
+	}
+
+	private readBodyAsBytes(body: unknown): { stream: ReadableStream<Uint8Array>; size: number } {
+		if (!body) {
+			return { stream: new ReadableStream<Uint8Array>(), size: 0 };
+		}
+
+		const uint8ArrayToStream = (uint8Array: Uint8Array): ReadableStream<Uint8Array> => {
+			return new ReadableStream({
+				start(controller) {
+					controller.enqueue(uint8Array);
+					controller.close();
+				},
+			});
+		};
+
+		if (typeof body === "string") {
+			return { stream: uint8ArrayToStream(new TextEncoder().encode(body)), size: body.length };
+		}
+		if (body instanceof Blob) {
+			return { stream: body.stream(), size: body.size };
+		}
+		if (body instanceof ArrayBuffer) {
+			return { stream: uint8ArrayToStream(new Uint8Array(body)), size: body.byteLength };
+		}
+		if (ArrayBuffer.isView(body)) {
+			return { stream: uint8ArrayToStream(new Uint8Array(body.buffer, body.byteOffset, body.byteLength)), size: body.byteLength };
+		}
+		throw new Error("Unsupported body type");
+	}
+
+	private async compressRequestBody(body: unknown): Promise<{
+		compressedBody: ArrayBuffer | Buffer;
+		size: number;
+	}> {
+		if (!inNodeEnv()) {
+			// in browser
+			const compressionData = this.readBodyAsBytes(body);
+			const compressedBody = await this.compressUsingCompressionStream(compressionData.stream);
+			return {
+				compressedBody: compressedBody.body,
+				size: compressedBody.size,
+			};
+		} else {
+			// In Node.js
+			const compressedBody = await this.compressUsingZlib(body);
+			return {
+				compressedBody,
+				size: compressedBody.length,
+			};
+		}
+	}
+
+	private async compressUsingZlib(body: unknown): Promise<Buffer> {
+		// @ts-ignore
+		const zlib = await import("zlib");
+		return new Promise((resolve, reject) => {
```

Comment:
We might need an additional await here for consistency, please double check

Reply:
@baywet could you clarify this abit

Reply:
we're returning the promise without awaiting it line 199. I think that TypeScript/JavaScript kind of "gets" what we're trying to do since promise chaining back in the days also accepted a returned promise. But it might be worth awaiting it as well to ensure things are properly chained since we're awaiting another promise above.